scripts/imgtool/main.py

```diff
@@ -480,10 +505,57 @@ def sign(key, public_key_format, align, version, pad_sig, header_size,
         }
 
     img.create(key, public_key_format, enckey, dependencies, boot_record,
-               custom_tlvs, int(encrypt_keylen), clear, baked_signature,
-               pub_key, vector_to_sign)
+               custom_tlvs, compression_tlvs, int(encrypt_keylen), clear,
+               baked_signature, pub_key, vector_to_sign)
+
+    if os.path.splitext(infile)[-1].lower() == '.bin' :
+        if compression == "lzma2" :
+            compressed_img = image.Image(version=decode_version(version),
+                      header_size=header_size, pad_header=pad_header,
+                      pad=pad, confirm=confirm, align=int(align),
+                      slot_size=slot_size, max_sectors=max_sectors,
+                      overwrite_only=overwrite_only, endian=endian,
+                      load_addr=load_addr, rom_fixed=rom_fixed,
+                      erased_val=erased_val, save_enctlv=save_enctlv,
+                      security_counter=security_counter, max_align=max_align)
+            compressed_file =  re.sub('bin$', 'compressed.bin', infile)
+            compression_filters = [
+                {"id": lzma.FILTER_LZMA2, "preset": comp_default_preset,
+                    "dict_size": comp_default_dictsize, "lp": comp_default_lp,
+                     "lc": comp_default_lc}
+            ]
+            in_fd=open(infile, "rb")
+            with lzma.open(compressed_file, "wb", filters=compression_filters,
+                format=lzma.FORMAT_RAW) as out_fd:
+                out_fd.write(in_fd.read())
+            in_fd.close()
+            uncompressed_size = os.path.getsize(infile)
+            compressed_size = os.path.getsize(compressed_file)
+            print("deflated", uncompressed_size, "=>", compressed_size)
+            compression_tlvs["DECOMP_SIZE"] = struct.pack(
+                img.get_struct_endian() + 'L', img.image_size)
+            compression_tlvs["DECOMP_SHA"] = img.image_hash
+            compression_tlvs["DECOMP_SIGNATURE"] = img.signature
+            compression_tlvs_size = len(compression_tlvs["DECOMP_SIZE"])
+            compression_tlvs_size += len(compression_tlvs["DECOMP_SHA"])
+            compression_tlvs_size += len(compression_tlvs["DECOMP_SIGNATURE"])
+            if (compressed_size + compression_tlvs_size) < uncompressed_size:
+                compression_header = create_lzma2_header(
+                    dictsize = comp_default_dictsize, pb = comp_default_pb,
+                    lc = comp_default_lc, lp = comp_default_lp)
+                compressed_img.load(compressed_file, compression_header)
+                compressed_img.create(key, public_key_format, enckey,
+                   dependencies, boot_record, custom_tlvs, compression_tlvs,
+                   compression, int(encrypt_keylen), clear, baked_signature,
+                   pub_key, vector_to_sign)
+                img = compressed_img
+            else:
+                print("Ommiting compression because it did nothing")
```

Comment:
it did nothing? Error should properly describe what is happening, if this prevents using compressed image if it is larger than the comment in the help text should be removed